archunit/src/main/java/com/tngtech/archunit/library/Architectures.java

```diff
@@ -289,4 +291,114 @@ public String toString() {
             }
         }
     }
+
+    @PublicAPI(usage = ACCESS)
+    public static OnionArchitecture onionArchitecture() {
+        return new OnionArchitecture();
+    }
+
+    public static final class OnionArchitecture implements ArchRule {
+        private static final String DOMAIN_MODEL_LAYER = "domain model";
+        private static final String DOMAIN_SERVICE_LAYER = "domain service";
+        private static final String APPLICATION_LAYER = "application";
+        private static final String ADAPTER_LAYER = "adapter";
+
+        private final Set<String> domainModelPackageIdentifiers = new LinkedHashSet<>();
+        private final Set<String> domainServicePackageIdentifiers = new LinkedHashSet<>();
+        private final Set<String> applicationPackageIdentifiers = new LinkedHashSet<>();
+        private final Map<String, Set<String>> adapterPackageIdentifiers = new LinkedHashMap<>();
+
+        private LayeredArchitecture layeredArchitectureDelegate = layeredArchitecture();
+
+        private OnionArchitecture() {
+        }
+
+        @PublicAPI(usage = ACCESS)
+        public OnionArchitecture domainModel(String... packageIdentifiers) {
+            domainModelPackageIdentifiers.addAll(asList(packageIdentifiers));
+            updateLayersDelegate();
+            return this;
+        }
+
+        @PublicAPI(usage = ACCESS)
+        public OnionArchitecture domainService(String... packageIdentifiers) {
+            domainServicePackageIdentifiers.addAll(asList(packageIdentifiers));
+            updateLayersDelegate();
+            return this;
+        }
+
+        @PublicAPI(usage = ACCESS)
+        public OnionArchitecture application(String... packageIdentifiers) {
+            applicationPackageIdentifiers.addAll(asList(packageIdentifiers));
+            updateLayersDelegate();
+            return this;
+        }
+
+        @PublicAPI(usage = ACCESS)
+        public OnionArchitecture adapter(String name, String... packageIdentifiers) {
+            adapterPackageIdentifiers.put(name, new LinkedHashSet<>(asList(packageIdentifiers)));
+            updateLayersDelegate();
+            return this;
+        }
+
+        private void updateLayersDelegate() {
+            layeredArchitectureDelegate = layeredArchitecture()
```

Comment:
Do you think this must be done after every change, or would once in the `check` method suffice?

Alternatively, you could initialize the `layeredArchitectureDelegate` in the constructor and hold the `LayerDefinition`s as fields. Then, in `domainModel()`, you'd only have to call `domainModelLayer.definedBy(packageIdentifiers)`.

Reply:
done :)